File: frontend/src/services/postService.ts
import apiClient from "../api/axiosInstance";
import { CreatePostPayload, Post, PostStatus, UpdatePostPayload } from "../types/post";

const postService = {
  create: async (payload: CreatePostPayload) => {
    const { data } = await apiClient.post<Post>("/api/v1/posts", payload);
    return data;
  },

  getAll: async () => {
    const { data } = await apiClient.get<Post[]>("/api/v1/posts");
    return data;
  },

  getPublished: async () => {
    const { data } = await apiClient.get<Post[]>("/api/v1/posts/published");
    return data;
  },

  getBySlug: async (slug: string) => {
    const { data } = await apiClient.get<Post>(`/api/v1/posts/slug/${slug}`);
    return data;
  },

  getById: async (postId: number) => {
    const { data } = await apiClient.get<Post>(`/api/v1/posts/${postId}`);
    return data;
  },

  getByAuthor: async (authorId: number) => {
    const { data } = await apiClient.get<Post[]>(`/api/v1/posts/author/${authorId}`);
    return data;
  },

  getByAuthorAndStatus: async (authorId: number, status: PostStatus) => {
    const { data } = await apiClient.get<Post[]>(`/api/v1/posts/author/${authorId}/status/${status}`);
    return data;
  },

  update: async (postId: number, payload: UpdatePostPayload) => {
    const { data } = await apiClient.put<Post>(`/api/v1/posts/${postId}`, payload);
    return data;
  },

  search: async (keyword: string) => {
    try {
      const { data } = await apiClient.get<Post[] | { results?: Post[] }>("/api/v1/posts/search", {
        params: { keyword }
      });

      if (Array.isArray(data)) {
        return data;
      }

      if (Array.isArray(data.results)) {
        return data.results;
      }
    } catch {
      // fallback below
    }

    const { data } = await apiClient.get<Post[]>("/api/v1/posts/search", {
      params: { keyword }
    });
    return data;
  },
  
  delete: async (postId: number) => {
    const { data } = await apiClient.delete(`/api/v1/posts/${postId}`);
    return data;
  },

  incrementViewCount: async (postId: number) => {
    const { data } = await apiClient.post(`/api/v1/posts/${postId}/view`);
    return data;
  },

  like: async (postId: number, userId: number) => {
    const response = await apiClient.post(`/api/v1/posts/${postId}/like`, null, {
      headers: { "X-User-Id": String(userId) }
    });
    return response.status;
  },

  unlike: async (postId: number, userId: number) => {
    const { data } = await apiClient.post(`/api/v1/posts/${postId}/unlike`, null, {
      headers: { "X-User-Id": String(userId) }
    });
    return data;
  }
};

export default postService;

File: post-service/post-service/src/main/java/com/inkwell/post/service/PostService.java
package com.inkwell.post.service;

import com.inkwell.post.dto.CreatePostRequest;
import com.inkwell.post.dto.PostDTO;
import com.inkwell.post.dto.UpdatePostRequest;
import com.inkwell.post.entity.PostStatus;

import java.util.List;

/**
 * Service interface for Post operations
 * Defines contract for post-related business logic
 */
public interface PostService {

    /**
     * Create a new post in DRAFT status
     *
     * @param createPostRequest the request containing post data
     * @return the created PostDTO
     */
    PostDTO createPost(CreatePostRequest createPostRequest);

    /**
     * Retrieve a post by ID
     *
     * @param postId the post ID
     * @return the PostDTO if found
     * @throws com.inkwell.post.exception.PostNotFoundException if not found
     */
    PostDTO getPostById(Integer postId);

    /**
     * Retrieve a post by its slug
     *
     * @param slug the post slug
     * @return the PostDTO if found
     * @throws com.inkwell.post.exception.PostNotFoundException if not found
     */
    PostDTO getPostBySlug(String slug);

    /**
     * Get all posts (paginated or not, depending on implementation)
     *
     * @return list of PostDTOs
     */
    List<PostDTO> getAllPosts();

    /**
     * Get published posts only
     *
     * @return list of published PostDTOs
     */
    List<PostDTO> getPublishedPosts();

    /**
     * Get all posts by a specific author
     *
     * @param authorId the author ID
     * @return list of PostDTOs by the author
     */
    List<PostDTO> getPostsByAuthor(Integer authorId);

    /**
     * Get all posts by a specific author with a given status
     *
     * @param authorId the author ID
     * @param status the post status
     * @return list of PostDTOs
     */
    List<PostDTO> getPostsByAuthorAndStatus(Integer authorId, PostStatus status);

    /**
     * Search posts by keyword
     *
     * @param keyword the search keyword
     * @return list of matching PostDTOs
     */
    List<PostDTO> searchPosts(String keyword);

    /**
     * Update an existing post
     *
     * @param postId the post ID
     * @param updatePostRequest the update request
     * @return the updated PostDTO
     * @throws com.inkwell.post.exception.PostNotFoundException if not found
     */
    PostDTO updatePost(Integer postId, UpdatePostRequest updatePostRequest);

    /**
     * Delete a post
     *
     * @param postId the post ID
     * @throws com.inkwell.post.exception.PostNotFoundException if not found
     */
    void deletePost(Integer postId);

    /**
     * Delete all posts owned by an author.
     *
     * @param authorId the author ID
     */
    void deletePostsByAuthor(Integer authorId);

    /**
     * Increment view count for a post (called when post is viewed)
     * Once per unique session.
     *
     * @param postId the post ID
     * @param sessionId the unique session ID of the viewer
     */
    void incrementViewCount(Integer postId, String sessionId);

    /**
     * Increment likes count for a post
     *
     * @param postId the post ID
     */
    void incrementLikesCount(Integer postId);

    boolean likePost(Integer postId, Integer userId);

    boolean unlikePost(Integer postId, Integer userId);

    /**
     * Decrement likes count for a post (when user unlikes)
     *
     * @param postId the post ID
     */
    void decrementLikesCount(Integer postId);

    /**
     * Publish a post (change status from DRAFT to PUBLISHED)
     *
     * @param postId the post ID
     * @return the updated PostDTO
     * @throws com.inkwell.post.exception.PostNotFoundException if not found
     */
    PostDTO publishPost(Integer postId);

    /**
     * Unpublish a post
     *
     * @param postId the post ID
     * @return the updated PostDTO
     * @throws com.inkwell.post.exception.PostNotFoundException if not found
     */
    PostDTO unpublishPost(Integer postId);

    /**
     * Feature (pin) or unfeature a post
     *
     * @param postId the post ID
     * @param isFeatured whether to feature the post
     * @return the updated PostDTO
     */
    PostDTO featurePost(Integer postId, boolean isFeatured);

    List<PostDTO> getMostViewedPosts(int limit);

    List<com.inkwell.post.dto.AuthorAnalyticsDTO> getMostActiveAuthors(int limit);

    List<com.inkwell.post.dto.TagDTO> getTrendingTags(int limit);
}

